File: Javascript-DSA/Sorting/AlgoExpert/mergeSort/countInversions.js
function countInversions(array) {
    return countInversionsHelper(array, 0, array.length - 1);
}
function countInversionsHelper(array, startIdx, endIdx) {
    if(startIdx >= endIdx) return 0;
    let midIdx = Math.floor((startIdx + endIdx) / 2);
    let leftInversions = countInversionsHelper(array, startIdx, midIdx);
    let rightInversions = countInversionsHelper(array, midIdx + 1, endIdx);
    let mergedInversions = mergeAndCountInversions(array, startIdx, midIdx, midIdx + 1, endIdx);
    return leftInversions + rightInversions  + mergedInversions;
}
function mergeAndCountInversions(array, leftStartIdx, leftEndIdx, rightStartIdx, rightEndIdx) {
    let inversions = 0;
    let leftArrLength = leftEndIdx - leftStartIdx + 1;
    let rightArrLength = rightEndIdx - rightStartIdx + 1;

    let arr1 = new Array(leftArrLength);
    let arr2 = new Array(rightArrLength);

    for(let i = 0; i < leftArrLength; i++) {
        arr1[i] = array[leftStartIdx + i];
    }
    for(let i = 0; i < rightArrLength; i++) {
        arr2[i] = array[rightStartIdx + i];
    }

    let i = 0;
    let j = 0;
    let k = leftStartIdx;

    while(i < leftArrLength && j < rightArrLength) {
        if(arr1[i] <= arr2[j]) {
            array[k++] = arr1[i++];
        }else {
            inversions += leftArrLength - i;
            array[k++] = arr2[j++];
        }
    }

    while(i < leftArrLength) {
        array[k++] = arr1[i++];
    }
    while(j < rightArrLength) {
        array[k++] = arr2[j++];
    }
    
    return inversions;
}

let array = [2, 3, 3, 1, 9, 5, 6];

console.log(countInversions(array));

File: sorting/CountInversions.java
package Sorting.mergeSort;

public class CountInversions {
    public static void main(String[] args) {
        int[] array = {9, 8, 7, 6, 5, 4, 3, 2, 1, 0};
        long ans = countInversionsHelper(array, 0, array.length - 1);
        System.out.println(ans);
    }
    public static long countInversionsHelper(int[] array, int startIdx, int endIdx) {
        long mod=1000000007;
        if (startIdx == endIdx) return 0L;
        int midIdx = (int) Math.floor((int) (startIdx + (endIdx - startIdx) / 2) ) ;

        long leftInversions = countInversionsHelper(array, startIdx, midIdx) % mod;
        long rightInversions = countInversionsHelper(array, midIdx + 1, endIdx) % mod;
        long mergedInversions = doMergeAndCountInversions(array, startIdx, midIdx, midIdx+1, endIdx) % mod;
        return (leftInversions + rightInversions + mergedInversions) % mod;
    }
    public static long doMergeAndCountInversions (int[] array, int leftStartIdx, int leftEndIdx, int rightStartIdx, int rightEndIdx) {
        long inversionsCount = 0;
        int leftArrLength = leftEndIdx - leftStartIdx + 1;
        int rightArrLength = rightEndIdx - rightStartIdx + 1;

        int[] leftArr = new int[leftArrLength];
        int[] rightArr = new int[rightArrLength];

        System.arraycopy(array, leftStartIdx, leftArr, 0, leftArrLength);
        System.arraycopy(array, rightStartIdx, rightArr, 0, rightArrLength);

        int i = 0;
        int j = 0;
        int k = leftStartIdx;
        while (i < leftArrLength && j < rightArrLength) {
            if (leftArr[i] <= rightArr[j]) {
                array[k++] = leftArr[i++];
            }else {
                // count inversions
                inversionsCount += leftArrLength - i;
                // continue merge-algorithm
                array[k++] = rightArr[j++];
            }
        }

        while (i < leftArrLength) {
            array[k++] = leftArr[i++];
        }

        while (j < rightArrLength) {
            array[k++] = rightArr[j++];
        }

        return inversionsCount;
    }
}
